### DISCUSSION TITLE
Relative url

### DISCUSSION THREAD

**Discussion by koteswar-rao** [11ty/eleventy#2516] [Q&A]
How to make URL's be relative in 11ty?

**Comment by pdehaan** [comment]
I don't think there is an easy way to make links relative. There was a previous discussion in https://github.com/11ty/eleventy/issues/2099 but that ended up being very specific to a project and felt a bit complex.

Generally I just use [`pathPrefix`](https://www.11ty.dev/docs/config/#deploy-to-a-subdirectory-with-a-path-prefix) if I know I'll be deploying something to a subdirectory (ie: GitHub pages, etc):

> _"If your site lives in a different subdirectory (particularly useful with GitHub pages), use pathPrefix to specify this. It’s used by the [`url` filter](https://www.11ty.dev/docs/filters/url/) and inserted at the beginning of all absolute url href links. It does not affect your file structure. Leading or trailing slashes are all normalized away, so don’t worry about it."_

You can also set the `pathPrefix` in your .eleventy.js config file (versus via a command line arg), so it's possible to conditionally set that path prefix depending on something like an env var or `NODE_ENV` variable; so if you're doing a local build maybe `pathPrefix` is blank, or if you're doing a production build, it is set to wherever subdirectory you're deploying to.

```js
module.exports = function(eleventyConfig) {
  return {
    pathPrefix: "/eleventy-base-blog/"
  }
};
```

**Comment by samfeldstein** [comment]
I know this thread is old, but I was having the same issue, so dropping this here in case anybody stumbles on it, like me.

Assuming my problem is the same as @koteswar-rao's, adding the `<base>` element to the `<head>` seemed to solve this. Example: `<base href="http://localhost:8080/">`.

Relative links in the source folder (eg `../recipes/cake.md`) don't account for the build output's pretty link structure, which looks like this: `_site/recipes/cake/index.html`. The relative link can't account for that extra `cake/` folder.

I don't fully understand why, but adding that `<base>` link seems to take care of it.

I haven't tested this on a live site yet.

**Comment by forresto** [comment]
With Eleventy 3.0.0 I found a solution that makes everything relative, including internal links, images, head link tags... everything with `src` or `href`. And you don't have to change your templates or content.

I can open a built html file without a server (`file://`). And [Quick Look](https://xoxo.zone/@forresto/113913565699389935) works.

Should this be an official option? It would make github pages build not need a special [`--pathprefix`](https://github.com/11ty/eleventy-base-blog/blob/2819f9521e170ca37fbc34c130213cbdcff03e3a/package.json#L9) and you could move a built site to any directory without rebuilding. It makes the static site more portable. 

I can't find docs for `eleventyConfig.htmlTransformer.addUrlTransform` but it is used [under the hood of HtmlBasePlugin](https://github.com/11ty/eleventy/blob/main/src/Plugins/HtmlBasePlugin.js).

## _config/relative-links.js

``` js
/** Referring to HtmlBasePlugin.js
 *
 *  This plugin tries to make all URLs in the HTML output relative to the page.
 *
 *  Useful for:
 *  * browsing via file://
 *  * gh-pages in subdirectory repo
 *  * unsure where in the directory structure the site will be hosted
 *
 *  We're expecting the internal links to start with "/"
 *
 *  todo?
 *  * option to include "index.html" for those directory links, for extra file:// compat
 *
 */

import path from "path";

export default function (eleventyConfig) {
	// Apply to all HTML output in your project
	eleventyConfig.htmlTransformer.addUrlTransform(
		"html",
		function makeUrlRelative(urlInMarkup) {
			// Skip empty URLs, non-root-relative URLs, and dev server image transform URLs
			if (
				!urlInMarkup 
				|| !urlInMarkup.startsWith("/") 
				|| urlInMarkup.startsWith("/.11ty/") 
				|| urlInMarkup.startsWith("//")
			) {
				return urlInMarkup;
			}

			// Get base directory path (keep trailing slash for index pages)
			const fromDir = this.url.endsWith("/") ? this.url : path.dirname(this.url);

			let relativePath = path.relative(fromDir, urlInMarkup);

			// Add ./ for same-directory references
			if (!relativePath.startsWith(".")) {
				relativePath = "./" + relativePath;
			}

			// Preserve trailing slash from original URL
			if (urlInMarkup.endsWith("/") && !relativePath.endsWith("/")) {
				relativePath += "/";
			}

			// console.log(this.url, fromDir, urlInMarkup, relativePath);
			return relativePath;
		},
		{
			priority: -1, // run last last (after PathToUrl)
		},
	);
}
```

## eleventy.config.js

``` js
import relativeLinks from "./_config/relative-links.js";

export default async function (eleventyConfig) {
  // ... other config

  // At end
  eleventyConfig.addPlugin(relativeLinks);
}
```

**Reply by forresto** [reply]
I had a use case beyond file:// browsing. I needed to publish a subset of my site to a course site with [a bunch of subdirectories](https://fabacademy.org/2025/labs/aalto/students/forrest-oliphant/fab-academy/01-potential-final-project.html).

**Reply by forresto** [reply]
Even without HtmlBasePlugin all my links start with "/" on build.

Plus it's easy to write templates with eg `<a href="/" class="home-link">` and then let this transform take care of relativizing the link depending on where the html goes in the hierarchy.

**Reply by nycki93** [reply]
This is great! This means I can host my site on a subdomain like example.com/~nycki and then later add a "proper" domain without having to break everything! Honestly that should be the default!!
